src/python/pants/option/parser.py

```diff
@@ -221,13 +265,64 @@ def add_flag_val(v):
+    # See if there are any unconsumed flags remaining, and if so, raise a ParseError.
     if flag_value_map:
-      raise ParseError('Unrecognized command line flags on {}: {}'.format(
-        self._scope_str(), ', '.join(flag_value_map.keys())))
+      self._raise_error_for_invalid_flag_names(
+        flag_value_map,
+        all_scoped_flag_names=list(get_all_scoped_flag_names()),
+        levenshtein_max_distance=levenshtein_max_distance)
 
     return namespace
 
+  def _raise_error_for_invalid_flag_names(self, flag_value_map, all_scoped_flag_names,
+                                          levenshtein_max_distance):
+    """Identify similar option names to unconsumed flags and raise a ParseError with those names."""
+    matching_flags = {}
+    for flag_name in flag_value_map.keys():
+      # We will be matching option names without their leading hyphens, in order to capture both
+      # short and long-form options.
+      flag_minus_leading_dashes = re.sub(r'^-+', '', flag_name)
+      scoped_name = (
+        '{}-{}'.format(self.scope, flag_minus_leading_dashes)
+        if self.scope != GLOBAL_SCOPE
+        else flag_minus_leading_dashes)
+      suffix_matching_option_names = []
+      levenshtein_matching_option_names = defaultdict(list)
+      for other_name in all_scoped_flag_names:
+        other_name_no_dashes = re.sub(r'^-+', '', other_name)
+        # If the invalid option on the command line is the ending of another option, display
+        # that. This can occur when an option is used without the appropriate prefix scope.
+        if other_name_no_dashes.endswith(scoped_name):
+          suffix_matching_option_names.append(other_name)
+        else:
+          # If an option name is similar to the one entered, according to
+          # --option-name-check-distance, display that. This covers misspellings!
+          levenshtein_distance = Levenshtein.distance(scoped_name, other_name_no_dashes)
+          if levenshtein_distance <= levenshtein_max_distance:
+            levenshtein_matching_option_names[levenshtein_distance].append(other_name)
+      # If any option name's suffix matched the invalid flag, put those first. Then, display the
+      # option names matching via edit distance in a deterministic way.
+      all_matching_scoped_option_names = suffix_matching_option_names + [
+        flag
+        for distance in sorted(levenshtein_matching_option_names.keys())
+        for flag in sorted(levenshtein_matching_option_names[distance])
+      ]
+      if all_matching_scoped_option_names:
+        matching_flags[flag_name] = all_matching_scoped_option_names
+
+    if matching_flags:
+      suggestions_message = '\n{}'.format('\n'.join(
+        '{}: [{}]'.format(flag_name, ', '.join(matches))
+        for flag_name, matches in matching_flags.items()
+      ))
+    else:
+      suggestions_message = ' <none>'
+    raise ParseError(
+      'Unrecognized command line flags on {scope}: {flags}. Suggestions:{suggestions}'
```

Comment:
I think I'd rather strip the `Suggestions:` prefix if there are no suggestions, rather than printing `Suggestions: <none>`

Reply:
Done!